plugin/pkg/admission/podplacementrestriction/admission.go

```diff
@@ -0,0 +1,205 @@
+		return nil
+	}
+
+	obj := a.GetObject()
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		glog.Errorf("expected pod but got %s", a.GetKind().Kind)
+		return nil
+	}
+
+	if !p.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("not yet ready to handle request"))
+	}
+
+	nsName := a.GetNamespace()
+	namespace, err := p.namespaceLister.Get(nsName)
+	if errors.IsNotFound(err) {
+		// in case of latency in our caches, make a call direct to storage to verify that it truly exists or not
+		namespace, err = p.client.Core().Namespaces().Get(nsName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return err
+			}
+			return errors.NewInternalError(err)
+		}
+	} else if err != nil {
+		return errors.NewInternalError(err)
+	}
+
+	//namespace := namespaceObj.(*api.Namespace)
+	nsTolerations, err := p.getNamespaceTolerations(namespace, NSDefaultTolerations)
+	if err != nil {
+		return err
+	}
+
+	var finalTolerations []api.Toleration
+	if len(nsTolerations) > 0 {
+		if len(pod.Spec.Tolerations) > 0 {
+			if tolerations.IsConflict(nsTolerations, pod.Spec.Tolerations) {
+				return fmt.Errorf("namespace tolerations and pod tolerations conflict")
+			}
+
+			// modified pod tolerations = namespace tolerations + current pod tolerations
+			finalTolerations = tolerations.MergeTolerations(nsTolerations, pod.Spec.Tolerations)
+		} else {
+			finalTolerations = nsTolerations
+
+		}
+	} else {
+		finalTolerations = pod.Spec.Tolerations
+	}
+
+	if len(finalTolerations) > 0 {
+		whiteList, err := p.getNamespaceTolerations(namespace, NSWLTolerations)
+		if err != nil {
+			return err
+		}
+
+		if len(whiteList) > 0 {
+			// check if the merged pod tolerations satisfy its namespace whitelist
+			if !tolerations.VerifyAgainstWhitelist(finalTolerations, whiteList) {
+				return fmt.Errorf("pod tolerations (possibly merged with namespace default tolerations) conflict with its namespace whitelist")
+			}
+		}
+	}
+
+	pod.Spec.Tolerations = finalTolerations
+	return nil
+
+}
+
+func NewPodTolerationsPlugin(pluginConfig *pluginapi.Configuration) *podTolerationsPlugin {
+	return &podTolerationsPlugin{
+		Handler:      admission.NewHandler(admission.Create),
+		pluginConfig: pluginConfig,
+	}
+}
+
+func (a *podTolerationsPlugin) SetInternalClientSet(client clientset.Interface) {
+	a.client = client
+}
+
+func (p *podTolerationsPlugin) SetInformerFactory(f informers.SharedInformerFactory) {
+	namespaceInformer := f.Core().InternalVersion().Namespaces()
+	p.namespaceLister = namespaceInformer.Lister()
+	p.SetReadyFunc(namespaceInformer.Informer().HasSynced)
+
+}
+
+func (p *podTolerationsPlugin) Validate() error {
+	if p.namespaceLister == nil {
+		return fmt.Errorf("missing namespaceLister")
+	}
+	if p.client == nil {
+		return fmt.Errorf("missing client")
+	}
+	return nil
+}
+
+func (p *podTolerationsPlugin) getNamespaceTolerations(ns *api.Namespace, key string) ([]api.Toleration, error) {
```

Comment:
please move the ClusterDefault pieces into a separate operation so the code is clearer.

getNamespaceTolerations -- look at namespace only
getClusterTolerations -- look at cluster only